File: frontend/editor/src/core/utils/toolIOCompat.ts
/**
 * Whether a chain of steps can run, from the generated I/O table rather than by running it.
 *
 * Duplicated in the backend (`ToolChainValidator`) and the AI engine (`tool_io_compat.py`) on
 * purpose: this runs on every keystroke, and the desktop build has no cloud backend to ask.
 * `testing/tool-io-cases.json` pins all three to the same answers.
 */

import {
  TOOL_IO,
  toolIOFor,
  type ToolArity,
  type ToolFormat,
  type ToolIOSpec,
  type ToolIOTable,
} from "@app/types/toolIO";

export type ToolDiagnosticSeverity = "ERROR" | "WARN" | "INFO";

/** Stable identifiers so the UI can pick its own wording. */
export const ToolDiagnosticCode = {
  /** The step declares no I/O, so nothing past it can be checked. */
  Undeclared: "undeclared-operation",
  /** The previous step's output is not a format this step accepts. */
  FormatMismatch: "format-mismatch",
  /** The previous step's output depends on a parameter that is not set yet. */
  OutputUncertain: "output-uncertain",
  /** The pipeline's input files are not a format the first step accepts. */
  SourceMismatch: "source-mismatch",
  /** The previous step emits several files and this one runs once per file. */
  FanOut: "fan-out",
  /** The previous step emits several files and this one consumes them together. */
  FanIn: "fan-in",
} as const;

export type ToolDiagnosticCode =
  (typeof ToolDiagnosticCode)[keyof typeof ToolDiagnosticCode];

export interface ToolDiagnostic {
  stepIndex: number;
  severity: ToolDiagnosticSeverity;
  code: ToolDiagnosticCode;
  /** Interpolation values; the wording lives in the locale files. */
  detail: {
    operation: string;
    accepts?: ToolFormat[];
    produced?: ToolFormat;
  };
}

export interface ToolChainStep {
  /** A plain string: a stored pipeline may name an endpoint this build does not model. */
  operation: string;
  /** Only used to resolve an output that depends on one. */
  parameters?: Record<string, unknown>;
}

export interface ToolChainOptions {
  /** The format entering step one, when known. */
  sourceFormat?: ToolFormat;
  toolIO?: ToolIOTable;
}

interface ResolvedOutput {
  format: ToolFormat;
  arity: ToolArity;
  /** False when a conditional output turns on a parameter we cannot see. */
  certain: boolean;
}

function isMultiInput(arity: ToolArity): boolean {
  return arity === "MISO" || arity === "MIMO";
}

function isMultiOutput(arity: ToolArity): boolean {
  return arity === "SIMO" || arity === "MIMO";
}

function normalise(value: unknown): string {
  return value === null || value === undefined
    ? ""
    : String(value).trim().toLowerCase();
}

function acceptsFormat(spec: ToolIOSpec, format: ToolFormat): boolean {
  return (
    format === "ANY" ||
    spec.accepts.includes("ANY") ||
    spec.accepts.includes(format)
  );
}

/**
 * First case whose conditions all hold wins. If none match but one reads a parameter we cannot
 * see, the declared output comes back uncertain: an unseen value might have picked another branch.
 */
export function resolveOutput(
  spec: ToolIOSpec,
  parameters?: Record<string, unknown>,
): ResolvedOutput {
  let sawUnknownParam = false;
  for (const rule of spec.cases ?? []) {
    let allHold = true;
    for (const condition of rule.when) {
      if (!parameters || !(condition.param in parameters)) {
        sawUnknownParam = true;
        allHold = false;
        continue;
      }
      const value = normalise(parameters[condition.param]);
      allHold &&= condition.matches.some((match) => normalise(match) === value);
    }
    if (allHold) {
      return { format: rule.produces, arity: rule.arity, certain: true };
    }
  }
  return {
    format: spec.produces,
    arity: spec.arity,
    certain: !sawUnknownParam,
  };
}

export function validateToolChain(
  steps: ToolChainStep[],
  options: ToolChainOptions = {},
): ToolDiagnostic[] {
  const table = options.toolIO ?? TOOL_IO;
  const diagnostics: ToolDiagnostic[] = [];
  let carried: ResolvedOutput | null = null;

  steps.forEach((step, index) => {
    const spec = toolIOFor(step.operation, table);
    if (!spec) {
      diagnostics.push({
        stepIndex: index,
        severity: "WARN",
        code: ToolDiagnosticCode.Undeclared,
        detail: { operation: step.operation },
      });
      // Nothing is known past an undeclared step.
      carried = null;
      return;
    }

    // Only the first step is handed the pipeline's input. Every later step is handed the previous
    // step's output, which is simply unknown once an undeclared step intervened - checking it
    // against the input again would judge it on a format it never receives.
    if (index === 0) {
      if (options.sourceFormat && !acceptsFormat(spec, options.sourceFormat)) {
        diagnostics.push({
          stepIndex: index,
          severity: "ERROR",
          code: ToolDiagnosticCode.SourceMismatch,
          detail: {
            operation: step.operation,
            accepts: spec.accepts,
            produced: options.sourceFormat,
          },
        });
      }
    } else if (carried !== null) {
      diagnostics.push(...checkTransition(index, step, spec, carried));
    }
    carried = resolveOutput(spec, step.parameters);
  });

  return diagnostics;
}

function checkTransition(
  index: number,
  step: ToolChainStep,
  spec: ToolIOSpec,
  previous: ResolvedOutput,
): ToolDiagnostic[] {
  const detail = {
    operation: step.operation,
    accepts: spec.accepts,
    produced: previous.format,
  };

  if (previous.format === "NONE") {
    return [
      {
        stepIndex: index,
        severity: "ERROR",
        code: ToolDiagnosticCode.FormatMismatch,
        detail,
      },
    ];
  }

  if (!acceptsFormat(spec, previous.format)) {
    return [
      {
        stepIndex: index,
        // Unresolved output: may yet be fine once the step is configured.
        severity: previous.certain ? "ERROR" : "WARN",
        code: previous.certain
          ? ToolDiagnosticCode.FormatMismatch
          : ToolDiagnosticCode.OutputUncertain,
        detail,
      },
    ];
  }

  if (!previous.certain) {
    return [
      {
        stepIndex: index,
        severity: "WARN",
        code: ToolDiagnosticCode.OutputUncertain,
        detail,
      },
    ];
  }

  if (isMultiOutput(previous.arity)) {
    return [
      {
        stepIndex: index,
        severity: "INFO",
        code: isMultiInput(spec.arity)
          ? ToolDiagnosticCode.FanIn
          : ToolDiagnosticCode.FanOut,
        detail,
      },
    ];
  }

  return [];
}

/** What a newly appended step would be handed, or undefined when the last step declares nothing. */
export function chainOutputFormat(
  steps: ToolChainStep[],
  toolIO: ToolIOTable = TOOL_IO,
): ToolFormat | undefined {
  const last = steps.at(-1);
  if (!last) return undefined;
  const spec = toolIOFor(last.operation, toolIO);
  return spec ? resolveOutput(spec, last.parameters).format : undefined;
}

/** Unknown operations are not claimed to be a problem. */
export function toolAcceptsFormat(
  operation: string,
  format: ToolFormat,
  toolIO: ToolIOTable = TOOL_IO,
): boolean {
  const spec = toolIOFor(operation, toolIO);
  return spec ? acceptsFormat(spec, format) : true;
}

export function hasBlockingDiagnostics(diagnostics: ToolDiagnostic[]): boolean {
  return diagnostics.some((d) => d.severity === "ERROR");
}

export function diagnosticsForStep(
  diagnostics: ToolDiagnostic[],
  stepIndex: number,
): ToolDiagnostic[] {
  return diagnostics.filter((d) => d.stepIndex === stepIndex);
}

File: engine/src/stirling/services/tool_io_compat.py
"""Whether a chain of tool steps can actually run, decided from the generated tool I/O table
rather than by running it.

The same rules exist in the backend (``ToolChainValidator``) and the frontend
(``utils/toolIoCompat.ts``). They are duplicated on purpose: the frontend checks a pipeline on
every keystroke, and the engine checks a plan before spending a call per step filling in its
parameters. ``testing/tool-io-cases.json`` runs against all three, so a rule that changes in one
place and not the others fails there.
"""

from __future__ import annotations

from enum import StrEnum

from pydantic import Field

from stirling.models.base import ApiModel
from stirling.models.tool_io import TOOL_IO, ToolArity, ToolFormat, ToolIOSpec
from stirling.models.tool_models import ToolEndpoint


class DiagnosticSeverity(StrEnum):
    """Only ERROR means the chain cannot run as configured."""

    ERROR = "ERROR"
    WARN = "WARN"
    INFO = "INFO"


class DiagnosticCode(StrEnum):
    """Matching the backend and frontend implementations."""

    UNDECLARED = "undeclared-operation"
    FORMAT_MISMATCH = "format-mismatch"
    OUTPUT_UNCERTAIN = "output-uncertain"
    SOURCE_MISMATCH = "source-mismatch"
    FAN_OUT = "fan-out"
    FAN_IN = "fan-in"


class ToolDiagnostic(ApiModel):
    """One problem, against the step that cannot run."""

    step_index: int
    severity: DiagnosticSeverity
    code: DiagnosticCode
    message: str


class ToolChainStep(ApiModel):
    """``parameters`` is only used to resolve an output that depends on one."""

    operation: ToolEndpoint
    parameters: dict[str, object] = Field(default_factory=dict)


class ResolvedOutput(ApiModel):
    """``certain`` is false when a case reads a parameter we cannot see."""

    format: ToolFormat
    arity: ToolArity
    certain: bool


def _is_multi_input(arity: ToolArity) -> bool:
    return arity in (ToolArity.MISO, ToolArity.MIMO)


def _is_multi_output(arity: ToolArity) -> bool:
    return arity in (ToolArity.SIMO, ToolArity.MIMO)


def _accepts_format(spec: ToolIOSpec, candidate: ToolFormat) -> bool:
    return candidate == ToolFormat.ANY or ToolFormat.ANY in spec.accepts or candidate in spec.accepts


def _normalise(value: object) -> str:
    """Both sides of a condition go through this, so a declaration's casing cannot matter."""
    return "" if value is None else str(value).strip().lower()


def resolve_output(spec: ToolIOSpec, parameters: dict[str, object] | None) -> ResolvedOutput:
    """First case whose conditions all hold wins.

    If none match but one reads a parameter we cannot see, the declared output comes back
    uncertain: an unseen value might have picked another branch.
    """
    saw_unknown_param = False
    for rule in spec.cases:
        all_hold = True
        for condition in rule.when:
            if parameters is not None and condition.param in parameters:
                raw: object = parameters[condition.param]
            elif condition.default is not None:
                # The caller omitted it, so it takes the endpoint's default.
                raw = condition.default
            else:
                saw_unknown_param = True
                all_hold = False
                continue
            normalised = _normalise(raw)
            all_hold = all_hold and any(_normalise(m) == normalised for m in condition.matches)
        if all_hold:
            return ResolvedOutput(format=rule.produces, arity=rule.arity, certain=True)
    return ResolvedOutput(format=spec.produces, arity=spec.arity, certain=not saw_unknown_param)


def validate_tool_chain(
    steps: list[ToolChainStep],
    *,
    source_format: ToolFormat | None = None,
    tool_io: dict[ToolEndpoint, ToolIOSpec] | None = None,
) -> list[ToolDiagnostic]:
    """Every problem with the chain, against the step that cannot run."""
    table = TOOL_IO if tool_io is None else tool_io
    diagnostics: list[ToolDiagnostic] = []
    carried: ResolvedOutput | None = None

    for index, step in enumerate(steps):
        spec = table.get(step.operation)
        if spec is None:
            diagnostics.append(
                ToolDiagnostic(
                    step_index=index,
                    severity=DiagnosticSeverity.WARN,
                    code=DiagnosticCode.UNDECLARED,
                    message=(
                        f"{step.operation} does not declare what it accepts or produces, "
                        "so the rest of the chain cannot be checked."
                    ),
                )
            )
            # Nothing is known past an undeclared step.
            carried = None
            continue

        # Only the first step is handed the pipeline's input. Every later step is handed the
        # previous step's output, which is simply unknown once an undeclared step intervened -
        # checking it against the input again would judge it on a format it never receives.
        if index == 0:
            if source_format is not None and not _accepts_format(spec, source_format):
                diagnostics.append(
                    ToolDiagnostic(
                        step_index=index,
                        severity=DiagnosticSeverity.ERROR,
                        code=DiagnosticCode.SOURCE_MISMATCH,
                        message=(f"{step.operation} accepts {_describe(spec)} but the input is {source_format}."),
                    )
                )
        elif carried is not None:
            diagnostics.extend(_check_transition(index, step, spec, carried))
        carried = resolve_output(spec, step.parameters or None)

    return diagnostics


def _check_transition(index: int, step: ToolChainStep, spec: ToolIOSpec, previous: ResolvedOutput) -> list[ToolDiagnostic]:
    if previous.format == ToolFormat.NONE:
        return [
            ToolDiagnostic(
                step_index=index,
                severity=DiagnosticSeverity.ERROR,
                code=DiagnosticCode.FORMAT_MISMATCH,
                message=(f"The previous step returns a report rather than a file, so {step.operation} has nothing to run on."),
            )
        ]

    if not _accepts_format(spec, previous.format):
        message = f"{step.operation} accepts {_describe(spec)} but the previous step produces {previous.format}."
        # Unresolved output: may yet be fine once the step is configured.
        return [
            ToolDiagnostic(
                step_index=index,
                severity=DiagnosticSeverity.ERROR if previous.certain else DiagnosticSeverity.WARN,
                code=DiagnosticCode.FORMAT_MISMATCH if previous.certain else DiagnosticCode.OUTPUT_UNCERTAIN,
                message=message,
            )
        ]

    if not previous.certain:
        return [
            ToolDiagnostic(
                step_index=index,
                severity=DiagnosticSeverity.WARN,
                code=DiagnosticCode.OUTPUT_UNCERTAIN,
                message=(
                    f"The previous step's output depends on how it is configured, so {step.operation} may not be able to run."
                ),
            )
        ]

    if _is_multi_output(previous.arity):
        fan_in = _is_multi_input(spec.arity)
        return [
            ToolDiagnostic(
                step_index=index,
                severity=DiagnosticSeverity.INFO,
                code=DiagnosticCode.FAN_IN if fan_in else DiagnosticCode.FAN_OUT,
                message=(
                    f"{step.operation} combines every file the previous step produced."
                    if fan_in
                    else f"{step.operation} runs once for each file the previous step produced."
                ),
            )
        ]

    return []


def _describe(spec: ToolIOSpec) -> str:
    return " or ".join(sorted(f.value for f in spec.accepts))


def blocking(diagnostics: list[ToolDiagnostic]) -> list[ToolDiagnostic]:
    """The diagnostics that mean the chain cannot run."""
    return [d for d in diagnostics if d.severity == DiagnosticSeverity.ERROR]
